### PR TITLE
Predicatedpcj

### PR DESCRIPTION
**Pull request by noodlemoodle**
update of old foam2 pr to foam3 https://github.com/foam-framework/foam2/pull/4950

- add predicate fo capabilitycapabilityjunction
- add predicatedprerequisitecapabilityjunctiondao that filters the dao by predicate for non-admin users
- cache the pcjs map from crunchservice in the session and only use it when context user/realuser matches session user/agent

- updates on nanopay side : https://github.com/nanoPayinc/NANOPAY/pull/12122
- merge capabilities for signingofficer (US, CA, BR) : https://github.com/nanoPayinc/NANOPAY/pull/12123

### CODE REVIEW

File: src/foam/nanos/crunch/lite/CapablePayloadStatusDAO.js [hunk 1]
```diff
@@ -98,7 +98,7 @@ foam.CLASS({
           payload.setStatus(newStatus);
           // TODO Maybe use projection MLang
           var crunchService = (CrunchService) x.get("crunchService");
-          var depIds = crunchService.getDependentIds(x, payload.getCapability());
+          var depIds = crunchService.getDependentIds(x, payload.getCapability()); // todo ruby look into seems fine
```
Comment (noodlemoodle) [inline src/foam/nanos/crunch/lite/CapablePayloadStatusDAO.js:101]:
```suggestion
          var depIds = crunchService.getDependentIds(x, payload.getCapability());

File: src/foam/nanos/crunch/lite/ruler/CapableStatusChangeAdjustApprovalsRuleAction.js [hunk 2]
```diff
@@ -93,7 +93,7 @@ foam.CLASS({
               var crunchService = (CrunchService) x.get("crunchService");
               var payloadDAO = (DAO) capableNewObj.getCapablePayloadDAO(x);
 
-              List<String> prereqIdsList = crunchService.getPrereqs(newCapablePayload.getCapability());
+              List<String> prereqIdsList = crunchService.getPrereqs(x, newCapablePayload.getCapability()); // todo ruby getprereqforcapable???
```
Comment (noodlemoodle) [inline src/foam/nanos/crunch/lite/ruler/CapableStatusChangeAdjustApprovalsRuleAction.js:96]:
```suggestion
              List<String> prereqIdsList = crunchService.getPrereqs(x, newCapablePayload.getCapability());
```

File: src/foam/nanos/crunch/PredicatedPrerequisiteCapabilityJunctionDAO.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/nanos/crunch/PredicatedPrerequisiteCapabilityJunctionDAO.js:3]: 2021

Comment (noodlemoodle) [inline src/foam/nanos/crunch/PredicatedPrerequisiteCapabilityJunctionDAO.js:3]:
```suggestion
 * Copyright 2021 The FOAM Authors. All Rights Reserved.
```

**Review by Nauna — APPROVED** [review]
I believe this makes sense, also would like to merge and run. We should do this when we don't have a deployment planned - that day.

Also just curious about the subject being put into the context ... need to think this one through - would like to bring it up in R&D brain storming

File: src/foam/nanos/crunch/Capability.js [hunk 4]
```diff
@@ -392,7 +419,7 @@ foam.CLASS({
           subject.setUser((User) userDAO.find(ucj.getSourceId()));
         }
 
-        var prereqs = crunchService.getPrereqs(getId());
+        var prereqs = crunchService.getPrereqs(x.put("subject", ucj.getSubject(x)), getId());
```
Comment (jlhughes) [inline src/foam/nanos/crunch/Capability.js:422]: Thought, 
crunchService.getPrereqs(x, ucj, getId())  and the CrunchService can setup the subject.

**Pull request merged**